src/app/main.rs

```diff
@@ -506,11 +508,20 @@ where
 
             // Routes requests to their original destination endpoints. Used as
             // a fallback when service discovery has no endpoints for a destination.
+            //
+            // If the `l5d-force-id` header is present, then that identity is
+            // used as the server name when connecting to the endpoint.
             let orig_dst_router = svc::builder()
```

Comment:
By adding this logic only to the orig_dst_router, I think the `l5d-force-id` header can only be used for non-Kubernetes services.  This is the case for tap requests, but I would have expected the `l5d-force-id` functionality to work for any outbound request, including ones to Kubernetes services.

Reply:
This is an interesting wrinkle... I think we'd probably want to preemptively fail the request in the case that the force-id doesn't match the discovered id?

Reply:
The header logic is here because my understanding is we can only inspect request headers in a router. Because of this, I don't think we can check for `l5d-force-id` any lower (? - closer to the endpoint) in the stack. It is an interesting point; the current change is very specific to tap requests but not helpful for other use cases.

I could look into checking for this header in an earlier router and setting something on the request that could be passed down through the stack? I'll look into some different approaches to take and update this conversation.

Reply:
I'm going to move the header check up into the destination router. We can then check the `l5d-force-id` value against the discovered endpoint identity and error if they are not equal. This will also allow more general use cases of the header instead of it's current location in the `original_dst_router`